Mergin/plugin.py

```diff
@@ -243,41 +260,78 @@ def sync_project(self):
+            return
+
+        dlg = SyncDialog()
+        dlg.pull_start(self.mc, self.path, self.project_name)
+
+        dlg.exec_()  # blocks until success, failure or cancellation
+
+        if dlg.exception:
+            # pull failed for some reason
+            if isinstance(dlg.exception, LoginError):
+                self._login_error_message(dlg.exception)
+            elif isinstance(dlg.exception, ClientError):
+                QMessageBox.critical(self, "Project sync", "Client error: " + str(dlg.exception))
+            else:
+                # TODO: this should not happen normally - "known" exceptions should be recognized
+                msg = "Failed to download changes!\n\n{}".format(str(dlg.exception))
+                QMessageBox.critical(None, 'Project sync', msg, QMessageBox.Close)
+            return
+
+        if dlg.pull_conflicts:
+            msg = "Following conflicts between local and server version found: \n\n"
+            for item in dlg.pull_conflicts:
+                msg += item + "\n"
+            msg += "\nYou may want to fix them before upload otherwise they will be uploaded as new files. " \
+                   "Do you wish to proceed?"
+            btn_reply = QMessageBox.question(None, 'Conflicts found', msg,
+                                             QMessageBox.Yes | QMessageBox.No, QMessageBox.No)
+            if btn_reply == QMessageBox.No:
+                QApplication.restoreOverrideCursor()
                 return
 
-            QApplication.setOverrideCursor(Qt.WaitCursor)
-            conflicts = self.mc.pull_project(self.path)
-            if conflicts:
-                msg = "Following conflicts between local and server version found: \n\n"
-                for item in conflicts:
-                    msg += item + "\n"
-                msg += "\nYou may want to fix them before upload otherwise they will be uploaded as new files. " \
-                       "Do you wish to proceed?"
-                btn_reply = QMessageBox.question(None, 'Conflicts found', msg,
-                                                 QMessageBox.Yes | QMessageBox.No, QMessageBox.No)
-                if btn_reply == QMessageBox.No:
-                    QApplication.restoreOverrideCursor()
-                    return    
-            
-            self.mc.push_project(self.path)
-
-            qgis_files = find_qgis_files(self.path)
-            if QgsProject.instance().fileName() in qgis_files:
-                QgsProject.instance().read()
-
-            QApplication.restoreOverrideCursor()
+        if not dlg.is_complete:
+            # we were cancelled
+            return
+
+        ## pull finished, start push
+
+        dlg = SyncDialog()
+        dlg.push_start(self.mc, self.path, self.project_name)
+
+        dlg.exec_()  # blocks until success, failure or cancellation
+
+        self._reload_project() # TODO: only reload project if we pulled a newer version
```

Comment:
we can do it based on status, we can either skip some stages if not needed

Reply:
I would leave it as-is for the time being, as we will be changing the logic when working on https://github.com/lutraconsulting/mergin-py-client/issues/30

Reply:
ok

Reply:
hm, now when I tried it without files to download, first sync window pop up and immediately disappeared so it results in effect of blinking, maybe it would be better to skip so it is not confusing. 

But not a hard requirement, we can postpone.